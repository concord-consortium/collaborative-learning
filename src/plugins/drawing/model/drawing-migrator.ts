import { types } from "mobx-state-tree";
import { clone } from "lodash";
import { DrawingContentModel } from "./drawing-content";
import { playbackChanges } from "./drawing-change-playback";
import { kDrawingStateVersion } from "./drawing-types";

export const isDrawingChangesFormat = (snapshot: any) =>
              snapshot.changes && Array.isArray(snapshot.changes);

// When this model is instantiated it will actually modify DrawingContentModel
// This is unexpected and is demonstrated here:
// `src/models/mst.test.ts`
export const DrawingMigrator = types.snapshotProcessor(DrawingContentModel, {
  preProcessor(snapshot: any) {
    // Right now we have 4 kinds of potential input:
    // - Version 1.1.0: current format of state
    // - Version 1.0.0: needs groups migration
    // - No version and with `objects`, not `changes`: this is the old import format
    // - No version and with a `changes` array: this is the old format of state


    // Current format
    if(snapshot.version === kDrawingStateVersion) {
      return snapshot;
    }

    if(isDrawingChangesFormat(snapshot)) {
      return playbackChanges(snapshot.changes);
    }

    if (!("version" in snapshot) || snapshot.version === "1.0.0") {
      // Unversioned content with objects, not changes, should be compatible with version 1.0.0.

      // The format change in version 1.1.0 is that groups now have height and width properties,
      // and the contained objects positions and dimensions are relative to the group.

      const migrated = clone(snapshot);
      if (!("objects" in migrated)) {
        migrated.objects = [];
      }
      migrated.objects.forEach((obj: any) => {
        if (obj.type === "group") {
          // We can't determine proper values for height and width until the models are instantiated,
          // so we set them to 0. The next step happens in `DrawingContentModel.afterCreate`.
          if ("objectExtents" in obj) {
            delete obj.objectExtents;
          }
          obj.width = 0;
          obj.height = 0;
        }
      });
      migrated.version = kDrawingStateVersion;
      return migrated;
    }

    console.warn("Drawing Tile could not be loaded", snapshot);
    return {};
  }

});
