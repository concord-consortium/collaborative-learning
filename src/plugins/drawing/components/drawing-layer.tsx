import React from "react";
import { reaction, IReactionDisposer } from "mobx";
import { clone, isAlive, getSnapshot } from "mobx-state-tree";
import { observer } from "mobx-react";
import { extractDragTileType, kDragTileContent } from "../../../components/tiles/tile-component";
import { DrawingContentModelType, DrawingObjectMove } from "../model/drawing-content";
import { ToolTileModelType } from "../../../models/tiles/tile-model";
import { safeJsonParse } from "../../../utilities/js-utils";
import { ImageContentSnapshotOutType } from "../../../models/tiles/image/image-content";
import { gImageMap } from "../../../models/image-map";
import { SelectionBox } from "./selection-box";
import { DrawingObjectSnapshotForAdd, DrawingObjectType, DrawingTool,
  HandleObjectHover,
  IDrawingLayer} from "../objects/drawing-object";
import { Point, ToolbarSettings } from "../model/drawing-basic-types";
import { getDrawingToolInfos, renderDrawingObject } from "./drawing-object-manager";
import { ImageObject } from "../objects/image";

const SELECTION_COLOR = "#777";
const HOVER_COLOR = "#bbdd00";
const SELECTION_BOX_PADDING = 10;

/**  ======= Drawing Layer ======= */
interface DrawingToolMap {
  [key: string]: DrawingTool | undefined;
}

interface DrawingLayerViewProps {
  model: ToolTileModelType;
  readOnly?: boolean;
  scale?: number;
  onSetCanAcceptDrop: (tileId?: string) => void;
  imageUrlToAdd?: string;
  setImageUrlToAdd?: (url: string) => void;
}

interface DrawingLayerViewState {
  toolbarSettings?: ToolbarSettings;
  currentDrawingObject: DrawingObjectType|null;
  selectedObjects: DrawingObjectType[];
  selectionBox: SelectionBox|null;
  hoverObject: DrawingObjectType|null;
  objectsBeingDragged: DrawingObjectType[];
}

@observer
export class DrawingLayerView extends React.Component<DrawingLayerViewProps, DrawingLayerViewState>
    implements IDrawingLayer {
  public currentTool: DrawingTool|null;
  public tools: DrawingToolMap;
  private svgRef: React.RefObject<any>|null;
  private setSvgRef: (element: any) => void;
  private _isMounted: boolean;
  private disposers: IReactionDisposer[];

  constructor(props: DrawingLayerViewProps) {
    super(props);

    this.state = {
      currentDrawingObject: null,
      selectionBox: null,
      selectedObjects: [],
      hoverObject: null,
      objectsBeingDragged: [],
    };

    this.tools = {};
    const drawingToolInfos = getDrawingToolInfos();
    drawingToolInfos.forEach(info => {
      if (info.toolClass) {
        this.tools[info.name] = new info.toolClass(this);
      }
    });

    this.currentTool = this.tools.select!;

    this.svgRef = null;
    this.setSvgRef = (element) => {
      this.svgRef = element;
    };
  }

  public componentDidMount() {
    this._isMounted = true;
    this.disposers = [];

    this.disposers.push(reaction(
      () => this.getContent().selectedButton,
      selectedButton => this.syncCurrentTool(selectedButton)
    ));

    // TODO: the list of selected objects is operated on directly by the model for example:
    // deleteSelectedObjects. So it is redundant to have the selection stored here in the state
    // too. There are still a few places working with this list of selected objects from the
    // state instead of the model.
    this.disposers.push(reaction(
      () => this.getContent().selectedIds,
      selectedIds => {
        const selectedObjects = selectedIds.map(
          id => this.getContent().objectMap[id]).filter(obj => !!obj) as DrawingObjectType[];
        this.setState({ selectedObjects });
      }
    ));

    this.disposers.push(reaction(
      () => this.getContent().toolbarSettings,
      settings => this.setCurrentToolSettings(settings)
    ));
  }

  public componentDidUpdate(prevProps: DrawingLayerViewProps, prevState: DrawingLayerViewState) {
    if (this.props.imageUrlToAdd) {
      this.addImage(this.props.imageUrlToAdd);
      this.props.setImageUrlToAdd?.("");
    }
  }

  public componentWillUnmount() {
    this.disposers.forEach(disposer => disposer());

    this._isMounted = false;
  }

  public syncCurrentTool(selectedButton: string) {
    const settings = this.getContent().toolbarSettings;
    const tool = this.tools[selectedButton];

    if (!tool) {
      console.warn("Unknown tool selected", selectedButton);
      return;
    }

    tool.setSettings(settings);
    this.setCurrentTool(tool);
  }

  public addNewDrawingObject(drawingObject: DrawingObjectSnapshotForAdd) {
    this.getContent().addObject(drawingObject);
  }

  public setSelectedObjects(selectedObjects: DrawingObjectType[]) {
    this.setState({selectionBox: null, selectedObjects});

    const drawingContent = this.props.model.content as DrawingContentModelType;
    const selectedObjectIds = selectedObjects.map(object => object.id || "");
    drawingContent.setSelection(selectedObjectIds);
  }

  public getSelectedObjects(): DrawingObjectType [] {
    return this.state.selectedObjects;
  }

  public setCurrentTool(tool: DrawingTool|null) {
    this.currentTool = tool;
  }

  public setCurrentToolSettings(settings: ToolbarSettings) {
    this.currentTool?.setSettings(settings);
  }

  public getCurrentStamp() {
    const drawingContent = this.props.model.content as DrawingContentModelType;
    return drawingContent.currentStamp;
  }

  public handleMouseDown = (e: React.MouseEvent<HTMLDivElement>) => {
    if (!this.props.readOnly && this.currentTool) {
      this.currentTool.handleMouseDown(e);
    }
  };

  public handleObjectClick = (e: MouseEvent|React.MouseEvent<any>, obj: DrawingObjectType) => {
    if (!this.props.readOnly && this.currentTool) {
      this.currentTool.handleObjectClick(e, obj);
    }
  };

  public handleObjectHover: HandleObjectHover = (e, obj, hovering) => {
    if (!this.props.readOnly && this.currentTool === this.tools.select) {
      this.setState({hoverObject: hovering ? obj : null});
    }
  };

  // handles dragging of selected/hovered objects
  // TODO: it seems this could be cleaned up. Keeping the state variable objectsBeingDragged
  // locally in this function and modifying the local variable is not safe.
  public handleSelectedObjectMouseDown = (e: React.MouseEvent<any>, obj: DrawingObjectType) => {
    if (this.props.readOnly || (this.currentTool !== this.tools.select)) return;
    let moved = false;
    const {selectedObjects, hoverObject } = this.state;
    let { objectsBeingDragged } = this.state;
    let objectsToInteract: DrawingObjectType[];
    let needToAddHoverToSelection = false;

    //If the object you are dragging is selected then the selection should not be cleared
    //and all objects should be moved.
    //If the object you are dragging was not selected then only then all of the other objects
    //should be deselected and just the object you are dragging should be selected.
    if (hoverObject && !selectedObjects.some(object => object.id === hoverObject.id)) {
      needToAddHoverToSelection = true;
      if (e.shiftKey || e.metaKey){
        objectsToInteract = [hoverObject, ...selectedObjects];
      }
      else {
        objectsToInteract = [hoverObject];
      }
    } else {
      objectsToInteract = selectedObjects;
    }

    const starting = this.getWorkspacePoint(e);
    if (!starting) return;

    e.stopPropagation();

    const handleMouseMove = (e2: MouseEvent) => {
      e2.preventDefault();
      e2.stopPropagation();

      const current = this.getWorkspacePoint(e2);
      if (!current) return;
      const dx = current.x - starting.x;
      const dy = current.y - starting.y;
      moved = moved || ((dx !== 0) || (dy !== 0));

      if (objectsBeingDragged.length === 0) {
        objectsBeingDragged = objectsToInteract.map(object => clone(object));
        this.setState( {objectsBeingDragged});
      }

      objectsBeingDragged.forEach((object, index) => {
        object.setPosition(objectsToInteract[index].x + dx, objectsToInteract[index].y + dy);
      });

      if (needToAddHoverToSelection) {
        // we delay until we confirm that the user is dragging the objects before adding the hover object
        // to the selection, to avoid messing with the click to select/deselect logic
        this.setSelectedObjects(objectsToInteract);
        // Note: the hoverObject could be kind of in a weird state here. It might
        // be both selected and hovered at the same time. However it is more
        // simple to keep the hoverObject independent of the selection. It just
        // represents the current object the mouse is over regardless of whether
        // it is selected or not.
        needToAddHoverToSelection = false;
      }
    };
    const handleMouseUp = (e2: MouseEvent) => {
      e2.preventDefault();
      e2.stopPropagation();
      window.removeEventListener("mousemove", handleMouseMove);
      window.removeEventListener("mouseup", handleMouseUp);
      if (moved) {
        const moves: DrawingObjectMove[] = objectsToInteract.map((object, index) => {
          const draggedObject = objectsBeingDragged[index];
          return {
            id: object.id,
            destination: {x: draggedObject.x, y: draggedObject.y}
          };
        });
        this.getContent().moveObjects(moves);
      }
      else {
        this.handleObjectClick(e2, obj);
      }
      this.setState({ objectsBeingDragged: []});
    };

    window.addEventListener("mousemove", handleMouseMove);
    window.addEventListener("mouseup", handleMouseUp);
  };

  public startSelectionBox(p: Point) {
    this.setState({selectionBox: new SelectionBox(p)});
  }

  public updateSelectionBox(p: Point) {
    const {selectionBox} = this.state;
    if (selectionBox) {
      selectionBox.update(p);
      this.setState({selectionBox});
    }
  }

  public endSelectionBox(addToSelectedObjects: boolean) {
    const {selectionBox} = this.state;
    if (selectionBox) {
      selectionBox.close();
      const selectedObjects: DrawingObjectType[] = addToSelectedObjects ? this.state.selectedObjects : [];
      this.forEachObject((object) => {
        if (object.inSelection(selectionBox)) {
          if (selectedObjects.indexOf(object) === -1) {
            selectedObjects.push(object);
          }
        }
      });
      this.setSelectedObjects(selectedObjects);
    }
  }

  // when we add text, this filter can be used with this.renderObjects((object) => object.type !== "text")
  public renderObjects(_filter: (object: DrawingObjectType) => boolean) {
    return this.getContent().objects.map((object) => {
      if (!object || !_filter(object)) {
        return null;
      }
      return renderDrawingObject(object, this.handleObjectHover, this.handleSelectedObjectMouseDown);
    });
  }

  public renderSelectedObjects(selectedObjects: DrawingObjectType[], color: string) {
    return selectedObjects.map((object, index) => {
      let {nw: {x: nwX, y: nwY}, se: {x: seX, y: seY}} = object.boundingBox;
      nwX -= SELECTION_BOX_PADDING;
      nwY -= SELECTION_BOX_PADDING;
      seX += SELECTION_BOX_PADDING;
      seY += SELECTION_BOX_PADDING;
      return <rect
                key={index}
                data-testid="selection-box"
                x={nwX}
                y={nwY}
                width={seX - nwX}
                height={seY - nwY}
                fill={color}
                fillOpacity="0"
                stroke={color}
                strokeWidth="1.5"
                strokeDasharray="10 5"
                pointerEvents={"none"}
               />;
    });
  }

  //we want to populate our objectsBeingDragged state array

  public setCurrentDrawingObject(object: DrawingObjectType | null) {
    this.setState({currentDrawingObject: object});
  }

  public getWorkspacePoint = (e: MouseEvent|React.MouseEvent<any>): Point|null => {
    if (this.svgRef) {
      const scale = this.props.scale || 1;
      const rect = ((this.svgRef as unknown) as Element).getBoundingClientRect();
      return {
        x: (e.clientX - rect.left) / scale,
        y: (e.clientY - rect.top) / scale
      };
    }
    return null;
  };

  public render() {
    const hoveringOverAlreadySelectedObject =
      this.state.hoverObject
        ? this.state.selectedObjects.indexOf(this.state.hoverObject) !== -1
        : false;

    const idsBeingDragged = this.state.objectsBeingDragged.map(object => object.id);
    const objectsToRenderSelected = this.state.objectsBeingDragged.length > 0 ?
      this.state.objectsBeingDragged : this.state.selectedObjects;

    return (
      <div className="drawing-layer"
          data-testid="drawing-layer"
          onMouseDown={this.handleMouseDown}
          onDragOver={this.handleDragOver}
          onDragLeave={this.handleDragLeave}
          onDrop={this.handleDrop} >

        <svg xmlnsXlink="http://www.w3.org/1999/xlink" width={1500} height={1500} ref={this.setSvgRef}>
          {this.renderObjects(object => object.type === "image" && !idsBeingDragged.includes(object.id))}
          {this.renderObjects(object => object.type !== "image" && !idsBeingDragged.includes(object.id))}
          {this.state.objectsBeingDragged.map((object) => renderDrawingObject(object))}
          {this.renderSelectedObjects(objectsToRenderSelected, SELECTION_COLOR)}
          {(this.state.hoverObject && !hoveringOverAlreadySelectedObject && isAlive(this.state.hoverObject))
            ? this.renderSelectedObjects([this.state.hoverObject], HOVER_COLOR)
            : null}
          {this.state.currentDrawingObject
            ? renderDrawingObject(this.state.currentDrawingObject)
            : null}
          {this.state.selectionBox ? this.state.selectionBox.render() : null}
        </svg>
      </div>
    );
  }

  private getContent() {
    return this.props.model.content as DrawingContentModelType;
  }

  private isAcceptableImageDrag = (e: React.DragEvent<HTMLDivElement>) => {
    const { readOnly } = this.props;
    const toolType = extractDragTileType(e.dataTransfer);
    // image drop area is central 80% in each dimension
    const kImgDropMarginPct = 0.1;
    if (!readOnly && (toolType === "image")) {
      const eltBounds = e.currentTarget.getBoundingClientRect();
      const kImgDropMarginX = eltBounds.width * kImgDropMarginPct;
      const kImgDropMarginY = eltBounds.height * kImgDropMarginPct;
      if ((e.clientX > eltBounds.left + kImgDropMarginX) &&
          (e.clientX < eltBounds.right - kImgDropMarginX) &&
          (e.clientY > eltBounds.top + kImgDropMarginY) &&
          (e.clientY < eltBounds.bottom - kImgDropMarginY)) {
        return true;
      }
    }
    return false;
  };

  private handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
    const isAcceptableDrag = this.isAcceptableImageDrag(e);
    // TODO: what is this method used for?
    this.props.onSetCanAcceptDrop(isAcceptableDrag ? this.props.model.id : undefined);
    if (isAcceptableDrag) {
      e.dataTransfer.dropEffect = "copy";
      e.preventDefault();
    }
  };

  private handleDragLeave = (e: React.DragEvent<HTMLDivElement>) => {
    this.props.onSetCanAcceptDrop();
  };

  private handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    if (this.isAcceptableImageDrag(e)) {
      const dragContent = e.dataTransfer.getData(kDragTileContent);
      const parsedContent = safeJsonParse(dragContent);
      if (parsedContent) {
        const droppedContent: ImageContentSnapshotOutType = parsedContent.content;
        const droppedUrl = droppedContent.url;
        if (droppedUrl) {
          this.handleImageDrop(droppedUrl);
        }
        e.preventDefault();
        e.stopPropagation();
      }
    }
  };

  private handleImageDrop(url: string) {
    this.props.setImageUrlToAdd?.(url);
  }

  private addImage(url: string) {
    gImageMap.getImage(url)
      .then(imageEntry => {
        if (!this._isMounted || !imageEntry.contentUrl) return;
        const image = ImageObject.create({
          // The contentUrl is used here because that is the normalized URL for this
          // image.
          url: imageEntry.contentUrl,
          // The filename is stored because the imageEntry is temporary and the filename
          // cannot be recreated just from the URL.  In this case, the imageEntry is being
          // used to transport this filename from the source of the drop.
          filename: imageEntry.filename,
          x: 0,
          y: 0,
          width: imageEntry.width!,
          height: imageEntry.height!
        });
        this.addNewDrawingObject(getSnapshot(image));
      });
  }

  private forEachObject(callback: (object: DrawingObjectType, key?: string) => void) {
    const {objects} = this.getContent();
    objects.forEach((object) => {
      if (object) {
        callback(object, object.id);
      }
    });
  }
}
