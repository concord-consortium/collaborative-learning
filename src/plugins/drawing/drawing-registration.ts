import { registerTileComponentInfo } from "../../models/tiles/tile-component-info";
import { registerTileContentInfo } from "../../models/tiles/tile-content-info";
import { DrawingContentModel, DrawingToolMetadataModel, defaultDrawingContent } from "./model/drawing-content";
import { kDrawingToolID, kDrawingDefaultHeight } from "./model/drawing-types";
import DrawingToolComponent from "./components/drawing-tile";
import DrawingToolIcon from "../../clue/assets/icons/draw-tool.svg";
import { DrawingMigrator } from "./model/drawing-migrator";

registerTileContentInfo({
  id: kDrawingToolID,
  titleBase: "Drawing",
  // TODO: maybe there is a better way to do this kind of casting?
  //   The issue is that modelClass prop has a type of `typeof TileContentModel`,
  //   That type is pretty restrictive and doesn't accommodate the return of
  //   types.snapshotProcessor. There might be a better way to get a
  //   typescript type for a MST "Class" which is less restrictive
  modelClass: DrawingMigrator as typeof DrawingContentModel,
  metadataClass: DrawingToolMetadataModel,
  defaultHeight: kDrawingDefaultHeight,
  exportNonDefaultHeight: true,
  defaultContent: defaultDrawingContent
});

registerTileComponentInfo({
  id: kDrawingToolID,
  Component: DrawingToolComponent,
  tileEltClass: "drawing-tool-tile",
  Icon: DrawingToolIcon
});
