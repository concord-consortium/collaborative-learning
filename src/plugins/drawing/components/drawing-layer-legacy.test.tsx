import { render, screen } from "@testing-library/react";
import React from "react";
import { ToolTileModel } from "../../../models/tiles/tile-model";
import { DrawingContentModelType } from "../model/drawing-content";
import { DrawingLayerView } from "./drawing-layer";
import { DrawingToolDeletion, DrawingToolMove } from "../model/drawing-types";
import { LineObjectSnapshot } from "../objects/line";
import { VectorObjectSnapshot } from "../objects/vector";
import { RectangleObjectSnapshot } from "../objects/rectangle";
import { EllipseObjectSnapshot } from "../objects/ellipse";
import { ImageObjectSnapshot } from "../objects/image";
import { DrawingMigrator } from "../model/drawing-migrator";

// The drawing tile needs to be registered so the ToolTileModel.create
// knows it is a supported tile type
import "../drawing-registration";

let content, drawingLayerProps, drawingLayer;

const getDrawingObject = (objectContent: DrawingContentModelType) => {
  drawingLayerProps = {
    model: ToolTileModel.create({content: objectContent}),
    onSetCanAcceptDrop: (tileId?: string) => {
      throw new Error("Function not implemented.");
    }
  };
  render(<DrawingLayerView {...drawingLayerProps} />);
  drawingLayer = screen.getByTestId("drawing-layer");
  return drawingLayer.firstChild;
};

describe("Drawing Layer Components", () => {
  describe("Freehand Line", () => {
    const lineData: LineObjectSnapshot = {
      type: "line",
      id: "123",
      x: 10, y: 10,
      deltaPoints: [{ dx: 1, dy: 1 }, { dx: 2, dy: 2 }],
      stroke: "#888888",
      strokeDashArray: "3,3",
      strokeWidth: 1
    };

    it("adds a freehand line", () => {
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: lineData})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a freehand line", () => {
      const moves: DrawingToolMove = [{ id: "123", destination: {x: 5, y: 5} }];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: lineData}),
        JSON.stringify({action: "move", data: moves})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a freehand line", () => {
      const deleteObject: DrawingToolDeletion = [ "123" ];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: lineData}),
        JSON.stringify({action: "delete", data: deleteObject})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Vector line", () => {
    const vectorData: VectorObjectSnapshot = {
      type: "vector",
      id: "234",
      x: 10, y: 10,
      dx: 10, dy: 10,
      stroke: "#888888",
      strokeDashArray: "3,3",
      strokeWidth: 1
    };
    it("adds a Vector line", () => {
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: vectorData})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a vector line", () => {
      const moves: DrawingToolMove = [{ id: "234", destination: {x: 5, y: 5} }];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: vectorData}),
        JSON.stringify({action: "move", data: moves})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a vector line", () => {
      const deleteObject: DrawingToolDeletion = [ "234" ];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: vectorData}),
        JSON.stringify({action: "delete", data: deleteObject})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Rectangle", () => {
    const rectData: RectangleObjectSnapshot = {
      type: "rectangle",
      id: "345",
      x: 10, y: 10,
      width: 10, height: 10,
      fill: "#cccccc",
      stroke: "#888888",
      strokeDashArray: "3,3",
      strokeWidth: 1
    };
    it("adds a Rectangle", () => {
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: rectData})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a rectangle", () => {
      const moves: DrawingToolMove = [{ id: "345", destination: {x: 5, y: 5} }];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: rectData}),
        JSON.stringify({action: "move", data: moves})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a rectangle", () => {
      const deleteObject: DrawingToolDeletion = [ "345" ];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: rectData}),
        JSON.stringify({action: "delete", data: deleteObject})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Ellipse", () => {
    const ellipseData: EllipseObjectSnapshot = {
      type: "ellipse",
      id: "456",
      x: 10, y: 10,
      rx: 10, ry: 10,
      fill: "#cccccc",
      stroke: "#888888",
      strokeDashArray: "3,3",
      strokeWidth: 1
    };
    it("adds a ellipse", () => {
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: ellipseData})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a ellipse", () => {
      const moves: DrawingToolMove = [{ id: "456", destination: {x: 5, y: 5} }];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: ellipseData}),
        JSON.stringify({action: "move", data: moves})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a ellipse", () => {
      const deleteObject: DrawingToolDeletion = [ "456" ];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: ellipseData}),
        JSON.stringify({action: "delete", data: deleteObject})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });

  describe("Image", () => {
    const imageData: ImageObjectSnapshot = {
      type: "image",
      id: "567",
      url: "my/image/url",
      x: 10, y: 10,
      width: 10, height: 10,
    };
    it("adds an image", () => {
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: imageData})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("moves a image", () => {
      const moves: DrawingToolMove = [{ id: "567", destination: {x: 5, y: 5} }];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: imageData}),
        JSON.stringify({action: "move", data: moves})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
    it("deletes a image", () => {
      const deleteObject: DrawingToolDeletion = [ "567" ];
      content = DrawingMigrator.create({changes:[
        JSON.stringify({action: "create", data: imageData}),
        JSON.stringify({action: "delete", data: deleteObject})
      ]});
      expect(getDrawingObject(content)).toMatchSnapshot();
    });
  });
});
