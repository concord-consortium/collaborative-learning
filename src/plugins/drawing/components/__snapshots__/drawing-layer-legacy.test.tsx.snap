// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawing Layer Components Ellipse adds a ellipse 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <ellipse
      cx="10"
      cy="10"
      fill="#cccccc"
      pointer-events="visible"
      rx="10"
      ry="10"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Ellipse deletes a ellipse 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  />
</svg>
`;

exports[`Drawing Layer Components Ellipse moves a ellipse 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <ellipse
      cx="5"
      cy="5"
      fill="#cccccc"
      pointer-events="visible"
      rx="10"
      ry="10"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Freehand Line adds a freehand line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <path
      d="M 10 10 l 1 1 l 2 2"
      fill="none"
      fill-rule="nonzero"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Freehand Line deletes a freehand line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  />
</svg>
`;

exports[`Drawing Layer Components Freehand Line moves a freehand line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <path
      d="M 5 5 l 1 1 l 2 2"
      fill="none"
      fill-rule="nonzero"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Image adds an image 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <image
      height="10"
      href="my/image/url"
      pointer-events="visible"
      preserveAspectRatio="none"
      width="10"
      x="10"
      y="10"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Image deletes a image 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  />
</svg>
`;

exports[`Drawing Layer Components Image moves a image 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <image
      height="10"
      href="my/image/url"
      pointer-events="visible"
      preserveAspectRatio="none"
      width="10"
      x="5"
      y="5"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Rectangle adds a Rectangle 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <rect
      class="rectangle"
      fill="#cccccc"
      height="10"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
      width="10"
      x="10"
      y="10"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Rectangle deletes a rectangle 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  />
</svg>
`;

exports[`Drawing Layer Components Rectangle moves a rectangle 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <rect
      class="rectangle"
      fill="#cccccc"
      height="10"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
      width="10"
      x="5"
      y="5"
    />
  </g>
</svg>
`;

exports[`Drawing Layer Components Vector line adds a Vector line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <g
      class="vector"
      fill="#888888"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    >
      <line
        x1="10"
        x2="20"
        y1="10"
        y2="20"
      />
    </g>
  </g>
</svg>
`;

exports[`Drawing Layer Components Vector line deletes a vector line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  />
</svg>
`;

exports[`Drawing Layer Components Vector line moves a vector line 1`] = `
<svg
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <g
    class="object-canvas"
    data-testid="drawing-layer-object-canvas"
    transform="translate(0, 0) scale(1)"
  >
    <g
      class="vector"
      fill="#888888"
      pointer-events="visible"
      stroke="#888888"
      stroke-dasharray=""
      stroke-width="1"
    >
      <line
        x1="5"
        x2="15"
        y1="5"
        y2="15"
      />
    </g>
  </g>
</svg>
`;
